cpp/src/hdbscan/detail/soft_clustering.cuh

```diff
@@ -84,38 +90,53 @@ void all_points_dist_membership_vector(const raft::handle_t& handle,
   raft::matrix::copyRows<value_t, value_idx, size_t>(
     X, n_exemplars, n, exemplars_dense.data(), exemplar_idx, n_exemplars, stream, true);
 
-  // compute the distances using raft API
-  rmm::device_uvector<value_t> dist(m * n_exemplars, stream);
+  // compute the number of batches based on the batch size
+  value_idx n_batches;
 
-  switch (metric) {
-    case raft::distance::DistanceType::L2SqrtExpanded:
-      raft::distance::
-        distance<raft::distance::DistanceType::L2SqrtExpanded, value_t, value_t, value_t, int>(
-          handle, X, exemplars_dense.data(), dist.data(), m, n_exemplars, n, true);
+  if (batch_size == 0) {
+    n_batches = 1;
+    batch_size = n_queries;
+  }
+  else {
+    n_batches = raft::ceildiv((int)n_queries, (int)batch_size);
+  }
+  for(value_idx bid = 0; bid < n_batches; bid++) {
+    value_idx samples_per_batch = min(batch_size, (int)n_queries - bid*batch_size);
+    value_idx batch_offset = bid * batch_size;
+    rmm::device_uvector<value_t> dist(samples_per_batch * n_exemplars, stream);
+
+     // compute the distances using raft API
+    switch (metric) {
+      case raft::distance::DistanceType::L2SqrtExpanded:
+        raft::distance::
+          distance<raft::distance::DistanceType::L2SqrtExpanded, value_t, value_t, value_t, int>(
+          handle, query + batch_offset * n, exemplars_dense.data(), dist.data(), samples_per_batch, n_exemplars, n, true);
       break;
     case raft::distance::DistanceType::L1:
       raft::distance::distance<raft::distance::DistanceType::L1, value_t, value_t, value_t, int>(
-        handle, X, exemplars_dense.data(), dist.data(), m, n_exemplars, n, true);
+        handle, query + batch_offset * n, exemplars_dense.data(), dist.data(), samples_per_batch, n_exemplars, n, true);
       break;
     case raft::distance::DistanceType::CosineExpanded:
       raft::distance::
         distance<raft::distance::DistanceType::CosineExpanded, value_t, value_t, value_t, int>(
-          handle, X, exemplars_dense.data(), dist.data(), m, n_exemplars, n, true);
+          handle, query + batch_offset * n, exemplars_dense.data(), dist.data(), samples_per_batch, n_exemplars, n, true);
```

Comment:
Very nice! This is a good solution to provide some significant savings in memory.